### PR TITLE
webgpu: clamp surface dimensions to max_texture_dimension_2d

### LINKED ISSUE #7819: WebGpu backend panics in Surface::configure
### What Operating System(s) are you seeing this problem on?

macOS

### Which Wayland compositor or X11 Window manager(s) are you using?

Using Aerospace on macos

#### Environment

| | |
|---|---|
| **wezterm** | `20260117-154428-05343b38` (nightly) |
| **OS** | macOS 26.5 / 26.4.1 (both affected) |
| **Hardware** | Mac16,8 — Apple M4 Pro, 48 GB |
| **Displays** | 2560×1440 external QHD + 3024×1964 built-in Retina |
| **Window manager** | [Aerospace](https://github.com/nikitabobko/AeroSpace) (tiling) |
| **front_end** | `"WebGpu"` (explicit in config) |

```lua
config.front_end = "WebGpu"
config.scrollback_lines = 3500
```

### WezTerm version

20260117-154428-05343b38

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

No, and I'll explain why below

### Describe the bug

WebGpu backend panics in `Surface::configure` when requested size exceeds 16384px max texture dim (macOS, multi-display / tiling WM)

#### Summary

On macOS with `front_end = "WebGpu"`, wezterm panics and aborts when a window resize event produces a `Surface` width or height larger than 16384 px. The panic occurs inside `WindowView::did_resize`, which is an Objective-C → Rust FFI callback. Rust cannot unwind across it, so the process aborts immediately (`panic_cannot_unwind`).

In my setup this is triggered every 8–9 days by [[Aerospace](https://github.com/nikitabobko/AeroSpace)](https://github.com/nikitabobko/AeroSpace) re-tiling windows across a dual-display layout where the computed geometry, under Retina scaling, exceeds the 16384 px wgpu limit.

#### Panic message (from `WEZTERM_LOG=info`)

```
thread 'main' panicked at vendor/wgpu-25.0.2/src/backend/wgpu_core.rs:3561:18:
wgpu error: Validation Error

Caused by:
  In Surface::configure
    `Surface` width and height must be within the maximum supported texture size.
    Requested was (19872, 2260), maximum extent for either dimension is 16384.

thread caused non-unwinding panic. aborting.
```

The width **19872 px** is consistent across all crashes — it corresponds to the two Retina displays combined under their scale factors. The height varies (2260 or 2820) depending on the workspace layout at the time.

#### Crash stack (relevant frames)

```
core::panicking::panic_cannot_unwind
window::os::macos::window::WindowView::did_resize        ← panic here, cannot unwind
  ← -[NSWindow _setFrameCommon:display:fromServer:]
  ← -[NSNotificationCenter postNotificationName:object:userInfo:]
window::os::macos::window::WindowInner::set_inner_size
window::os::macos::connection::Connection::with_window_inner::{{closure}}
async_task::raw::RawTask::run
window::spawn::SpawnQueue::run
```

#### Crash history

| Date | Uptime | Requested size | Wake time |
|---|---|---|---|
| 2026-05-12 | 8 days | 19872 × 2260 | 1 s |
| 2026-05-22 | 9 days | (no log — not via logging launcher) | 40 min |
| 2026-05-26 | 4 days | 19872 × 2260 | 1 s |
| 2026-05-27 | 21 h | 19872 × 2820 | 14 min |

Low wake time suggests the trigger is Aerospace re-tiling all windows on wake from sleep, which sends resize events to every window simultaneously.

#### Root cause

`WebGpuState::resize()` in `wezterm-gui/src/termwindow/webgpu.rs` already guards against zero-sized surfaces (issue #2881) but does not guard against the upper limit:

```rust
// current code (simplified):
config.width = dims.pixel_width as u32;
config.height = dims.pixel_height as u32;
if config.width > 0 && config.height > 0 {
    self.surface.configure(&self.device, &config);  // ← panics if > max_texture_dimension_2d
}
```

`Surface::configure` validates against `device.limits().max_texture_dimension_2d` (16384 on Apple Silicon) and raises a wgpu validation error, which Rust turns into a panic. Because the call originates from an Objective-C notification callback (`did_resize`), the panic cannot unwind and the process aborts.

#### Fix

Clamp both dimensions to `device.limits().max_texture_dimension_2d` before calling `Surface::configure`. I have a working patch with tests:

```rust
pub(crate) fn clamp_surface_dimensions(width: u32, height: u32, max: u32) -> (u32, u32) {
    let clamped_w = width.min(max);
    let clamped_h = height.min(max);
    if clamped_w != width || clamped_h != height {
        log::warn!(
            "Clamped surface size from {}x{} to {}x{} (max_texture_dimension_2d={})",
            width, height, clamped_w, clamped_h, max
        );
    }
    (clamped_w, clamped_h)
}
```

Applied in both `resize()` and the initial `SurfaceConfiguration` in `new_impl()`.

I'm happy to open a PR with this patch.

### To Reproduce

Tests that reproduce the exact crash dimensions and verify the fix:

```rust
#[test]
fn clamps_width_exceeding_max() {
    // Exact dimensions from crash: Aerospace spanning two Retina displays
    assert_eq!(clamp_surface_dimensions(19872, 2260, 16384), (16384, 2260));
}

#[test]
fn clamps_both_dimensions() {
    assert_eq!(clamp_surface_dimensions(20000, 20000, 16384), (16384, 16384));
}
```

Running these tests against the unfixed code (returning `(width, height)` unchanged) fails as expected, confirming the tests catch the regression.

### Configuration

```lua
config.front_end = "WebGpu"
config.scrollback_lines = 3500
```

### Expected Behavior

_No response_

### Logs

_No response_

#### Workarounds tried

- Switching `front_end` to `OpenGL`:  has a different fatal panic in `draw_rect` on this Mac, so not viable.
- Reducing `scrollback_lines`: unrelated to this crash.
- Upgrading from the stale Feb-2024 stable to the current nightly, fixed unrelated memory leaks but this panic persists.

### PR DESCRIPTION
**Pull request by GeckoSplinter**
Fix: https://github.com/wezterm/wezterm/issues/7819 

Surface::configure raises a wgpu validation error if either dimension exceeds the adapter's max_texture_dimension_2d (16384 on Apple Silicon). When this happens inside the did_resize Objective-C → Rust FFI callback, Rust cannot unwind and the process aborts.

This is reproducible on macOS with tiling window managers (e.g. Aerospace) that transiently compute window geometry spanning multiple Retina displays on wake from sleep or workspace re-tile. The requested width in all observed crashes was consistently 19872 px (two displays combined under their scale factors), exceeding the 16384 px GPU limit.

Guard both the initial SurfaceConfiguration in new_impl() and the live resize() path by clamping to device.limits().max_texture_dimension_2d, mirroring the existing zero-size guard added in #2881. The stored self.dimensions value is left unclamped so the resize dedup guard keeps comparing against the size the window actually requested; clamping is applied only to the values handed to Surface::configure.

Add unit tests covering the exact crash dimensions, boundary cases, and a degenerate max==0 adapter, including a regression test that fails against the unfixed code.

AI usage: Claude Opus 4.8

### CODE REVIEW

File: wezterm-gui/src/termwindow/webgpu.rs [hunk 1]
```diff
@@ -559,3 +571,76 @@ impl WebGpuState {
...
+    (clamped_w, clamped_h)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::clamp_surface_dimensions;
+
+    #[test]
+    fn no_clamp_when_within_limit() {
+        assert_eq!(clamp_surface_dimensions(1920, 1080, 16384), (1920, 1080));
```
Comment (bew) [inline wezterm-gui/src/termwindow/webgpu.rs]: could you move the 'magic value' for the max dimension as a constant in the test module? so it's more explicit when reading 🙏

File: wezterm-gui/src/termwindow/webgpu.rs [hunk 2]
```diff
@@ -559,3 +571,76 @@ impl WebGpuState {
         }
     }
 }
+
+/// Clamp a requested surface size to the adapter's maximum texture dimension.
+///
+/// `Surface::configure` raises a validation error (which becomes a fatal panic
+/// when called from an Objective-C → Rust FFI callback) if either dimension
+/// exceeds `max`. This is reachable on macOS with tiling window managers that
+/// transiently compute window geometry spanning multiple Retina displays.
```
Comment (bew) [inline wezterm-gui/src/termwindow/webgpu.rs]: I've rephrased the doc here to avoid assuming responsibility (the function technically does not require the max to be from the gpu adapter).
Focus on the general warning, THEN mention when it can happen, and mention the issue.
(note didn't check line lengths)

```suggestion
/// Clamp a requested surface size to a maximum.
/// (e.g. the GPU adapter's maximum texture dimension)
///
/// This is necessary before `Surface::configure` as it raises a validation error if either
/// dimension exceeds the GPU supported max dimension.
///
/// This can happen on macOS with tiling window managers that transiently compute
/// window geometry spanning multiple Retina displays.
///  (which becomes a fatal panic when called from an Objective-C → Rust FFI callback)
/// See https://github.com/wezterm/wezterm/issues/7819.
```

File: wezterm-gui/src/termwindow/webgpu.rs [hunk 3]
```diff
@@ -559,3 +571,76 @@ impl WebGpuState {
...
+fn clamp_surface_dimensions(width: u32, height: u32, max: u32) -> (u32, u32) {
```
Comment (bew) [inline wezterm-gui/src/termwindow/webgpu.rs]: please use the full `max_texture_dimension_2d` name for `max`, as you name it as such in the log below

**Review by bew — COMMENTED** [review]
Hello! The changes look good overall, thank you!

There was already a hard check when creating `WebGpuTexture` for the rendering pipeline, but none for the surface used for the window.

Just a few comments on phrasing/naming.

Also, how much AI and what tools/model was used for this PR ?

---

~~Note that WezTerm is currently on hold/unmaintained, Wez hasn't been responding to PRs for a while. 
👉 I'm interested in starting a community fork, see https://github.com/wezterm/wezterm/issues/6341#issuecomment-4483047032, so this might gets merged there.~~
FYI Development/Maintenance on Wezterm is slowly coming back on track! 💪 ([ref](https://github.com/wezterm/wezterm/issues/6341#issuecomment-4620396384))

**Comment by GeckoSplinter** [comment]
> Also, how much AI and what tools/model was used for this PR ?

I investigated the issue for few weeks with claude (Opus 4.8 recently and cross review with a second agent).
I have no experience in rust so sorry if the code is not perfect and I am just an SRE
But I try to review the code as much as possible 

I will take your comments into account 



> Note that WezTerm is currently on hold/unmaintained, Wez hasn't been responding to PRs for a while.
👉 I'm interested in starting a community fork, see https://github.com/wezterm/wezterm/issues/6341#issuecomment-4483047032, so this might gets merged there.
(you'll be notified if so) 

Sorry to hear that :(

**Comment by bew** [comment]
Hello!

FYI Development/Maintenance on Wezterm is slowly coming back on track, Wez being a bit more available and me having new github powers! 💪 ([ref](https://github.com/wezterm/wezterm/issues/6341#issuecomment-4620396384))

Once you've updated the PR I'll re-review again 🙏

**Comment by GeckoSplinter** [comment]
Hello @bew 

Sorry for the long time to reply on this.
I have updated the PR according to the review you have done 👍

File: wezterm-gui/src/termwindow/webgpu.rs [hunk 4]
```diff
@@ -604,32 +609,51 @@ fn clamp_surface_dimensions(width: u32, height: u32, max: u32) -> (u32, u32) {
 mod tests {
     use super::clamp_surface_dimensions;
 
+    /// Apple Silicon's `max_texture_dimension_2d`, and the limit hit in the
+    /// crash reports this fix addresses.
```
Comment (bew) [inline wezterm-gui/src/termwindow/webgpu.rs]: 'this fix' only makes sense in context.
Prefer to link to the issue instead

File: wezterm-gui/src/termwindow/webgpu.rs [hunk 5]
```diff
@@ -604,32 +609,51 @@ fn clamp_surface_dimensions(width: u32, height: u32, max: u32) -> (u32, u32) {
...
+    const MAX_TEXTURE_DIMENSION_2D: u32 = 16384;
```
Comment (bew) [inline wezterm-gui/src/termwindow/webgpu.rs]: This is not a generic value, let's rename with `SAMPLE_` prefix maybe?

**Review by bew — COMMENTED** [review]
Looking good otherwise, thank you!
